dds/DCPS/RTPS/Spdp.cpp

```diff
@@ -2822,7 +2831,12 @@ Spdp::SpdpTransport::handle_input(ACE_HANDLE h)
           plist[0]._d(PID_PARTICIPANT_GUID);
         }
 
-        outer_->data_received(data, plist, remote);
+        DCPS::WeakRcHandle<Spdp> wh(*outer_);
+        DCPS::RcHandle<Spdp> outer = wh.lock();
+
+        if (outer) {
```

Comment:
For WeakRcHandle to work in this way, the type of `outer_` needs to be changed to WeakRcHandle.  That would allow the WeakRcHandle to be initialized at a time when the RcObject is guaranteed to be valid.

Reply:
In general, I agree with what you're saying, but I think we can assume outer_ is still valid at this point. With the current design, as long as the reactor thread is still running (which it has to be, here, in handle_input() before data_received has a change to reverse lock anything), the object will still be valid.

I initially started by changing outer_ to a WeakRcHandle but it touches a lot of code with marginal advantage for this usage so I changed back and tried it this way. I haven't seen any complaints (for this section of code) from the address sanitizer so far, even running both reliable and durable versions of Thrasher in a loop.

I see the benefit for maintainability to move to WeakRcHandle, though, so if you still want me to make those changes let me know.

Reply:
It's fine to keep it that way, but add the info about the assumptions in a comment.

Reply:
The same operation needs / needed to happen in SpdpTransport, but because of the ChangeMulticastGroup jobs, which held references to SpdpTransport, the same guarantees couldn't be made. As a result, I opted for the full WeakRcHandle conversion (which touched a lot of lines but wasn't horrible) and removed these comments. The DpShutdown issues seem to be resolved on my windows build and Thrasher seems to run fine without ASAN issues. I'll kick off a new set of tests but this is probably ready for re-review (I'll mark it as such if I don't see anything major).